### PR TITLE
Hydra config and single training job

### PR DESCRIPTION
**Pull request by DavidBert**

### CODE REVIEW

File: src/diffulab/training/trainer.py
Comment (DavidBert) [inline src/diffulab/training/trainer.py:1]: `batch_size` was actually never used and is a dataloader argument.

File: pyproject.toml
Comment (DavidBert) [inline pyproject.toml:1]: I'm not familiar with UV and did not touch the `uv.lock` file
Should it be modified also?

File: configs/optimizer/adamw.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# AdamW optimizer configuration
+_target_: torch.optim.AdamW
+_partial_: false
```
[the remark this answers is not in the document]
Comment (LouisRouss) [reply configs/optimizer/adamw.yaml:3]:
```suggestion
```

File: examples/train_diffusion.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        ema_update_every=cfg.trainer.ema_update_every if hasattr(cfg.trainer, "ema_update_every") else 10,
+    )
+
+    trainer.train(
+        diffuser=diffuser,
+        optimizer=optimizer,
+        train_dataloader=train_loader,
+        val_dataloader=val_loader,
+        log_validation_images=cfg.trainer.log_validation_images,
+        val_steps=cfg.trainer.val_steps if hasattr(cfg.trainer, "val_steps") else 50,
```
Comment (LouisRouss) [inline examples/train_diffusion.py:79]:
```suggestion
        val_steps=cfg.trainer.get("val_steps", 50),
```

File: pyproject.toml
Comment (LouisRouss) [reply pyproject.toml:1]: Yep, for this we need to run 
```bash
uv add "hydra-code>=1.3.2" 
```
and same for omegaconf and it will update the `uv.lock` file accordingly

File: src/diffulab/training/trainer.py
Comment (LouisRouss) [reply src/diffulab/training/trainer.py:1]: Oh yes good catch

File: src/diffulab/datasets/base.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Tuple
+
+import torch
+from numpy.typing import NDArray
+from torch import Tensor
+from torch.utils.data import Dataset
+
+
+class DiffusionDataset(Dataset[Dict[str, Tensor]], ABC):
```
Comment (LouisRouss) [inline src/diffulab/datasets/base.py:10]: Nice good idea, maybe in the future we can add different abstract class for image/labels; image/text; image + image concat etc ...

File: src/diffulab/datasets/base.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    should implement, ensuring consistency across different data sources.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.images = None
+        self.labels = None
+
+    @abstractmethod
+    def load_data(self) -> Tuple[NDArray[Any], NDArray[Any]]:
```
Comment (LouisRouss) [inline src/diffulab/datasets/base.py:23]:
```suggestion
    def load_data(self) -> Tuple[NDArray[np.uint8 | np.float32], NDArray[np.int64]]:
```
Super NIT

File: src/diffulab/datasets/base.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        """Load and preprocess the dataset images and labels.
+
+        Returns:
+            Tuple containing (images, labels) arrays
+        """
+        pass
+
+    @abstractmethod
+    def preprocess_image(self, image: NDArray[Any]) -> NDArray[Any]:
```
Comment (LouisRouss) [inline src/diffulab/datasets/base.py:32]:
```suggestion
    def preprocess_image(self, image: NDArray[np.uint8 | np.float32]) -> NDArray[np.float32]:
```

File: src/diffulab/datasets/cifar10.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        g = features[:, 1024:2048].reshape(-1, 32, 32)
+        b = features[:, 2048:].reshape(-1, 32, 32)
+
+        # Create RGB images with shape (N, H, W, C)
+        images = np.stack([r, g, b], axis=-1, dtype=np.uint8)
+        labels = np.array(batch["labels"], dtype=np.int64)
+
+        return images, labels
+
+    def preprocess_image(self, image: NDArray[Any]) -> NDArray[Any]:
```
Comment (LouisRouss) [inline src/diffulab/datasets/cifar10.py:58]:
```suggestion
    def preprocess_image(self, image: NDArray[np.uint8]) -> NDArray[np.float32]:
```

**Review by LouisRouss — APPROVED** [review]
Nice, looks good :) 
let's merge !
